### ISSUE TITLE
Incorrect deserialisation of deeply nested PostgreSQL ARRAYs / UDTs

### ISSUE THREAD

**Issue by robertotru** [jOOQ/jOOQ#5613]
Hi folks

I was testing JOOQ master (3.9.0-dev) by trying to bind in _PostgreSQL 9.5_ something like:
`SELECT * FROM app_user_view WHERE uv_id=1;`

where I have defined

```
CREATE VIEW app_user_view AS
  SELECT
    u_id AS uv_user_id,
    ARRAY(
        SELECT ROW (
          g_id,
          group_roles :: role_type [],
        )::group_type
        FROM "app_group"
        WHERE ag_user_id = u_id
             AS uv_groups
  FROM "app_user";
```

From what you can see, If I try to bind the result of this `SELECT` with jOOQ, then jOOQ will need to return a single row (assuming that `u_id` is unique) such that it contains an array of UDTs. Each UDT will contain an array of UDTs as well (of type `role_type`, precisely).

Now, If I run the query with jOOQ master (1.3.9-dev), then I see that in class `DefaultBinding`, when the method `pgNewArray(Class<?> type, String string)` is called, the type `RoleTypeRecord` is correctly detected and assigned to `Class<?> component`. However, there is a problem that occurs when splitting the `string` value. Suppose that the `value` of `string`passed to `pgNewArray` is:

`String string = "{/"(1,ROLE_USER)/",/"(2,ROLE_ADMIN)/"}";`

Then it is done `String values = string.replaceAll("^\\{(.*)\\}$", "$1");` which correctly returns `"/"(1,ROLE_USER)/",/"(2,ROLE_ADMIN)/"";` but then the split `String[] split = values.split(",");` will ignore the parenthesis and return and array with 4 values and not with 2, that is the number of `String` objects to be returned to create the two UDTs. Clearly the 4 values are also wrong, because they will include the parenthesis, i.e. `(1`, `ROLE_USER)`, `(2`, and `ROLE_ADMIN)`.

Assuming this is an issue, please, let me know if I can help fixing this, e.g. by doing a better splitting  or whatever else you have in mind.

**Comment by lukaseder** [comment]
Thank you very much for reporting. I can reproduce a similar issue with the following isolated test case:

```sql
CREATE TYPE u_5613_1 AS (
  id   INT,
  name TEXT
);

CREATE TYPE u_5613_2 AS (
  id     INT,
  nested u_5613_1 []
);

CREATE VIEW v_5613 AS
SELECT
  1 AS id,
  ARRAY(
    SELECT ROW (
      2,
      null :: u_5613_1 []
    ) :: u_5613_2
    UNION ALL
    SELECT ROW (
      3,
      ARRAY[] :: u_5613_1 []
    ) :: u_5613_2
    UNION ALL
    SELECT ROW (
      4,
      ARRAY[ROW(1, 'A')] :: u_5613_1 []
    ) :: u_5613_2
    UNION ALL
    SELECT ROW (
      5,
      ARRAY[ROW(1, 'A'), ROW(2, 'B')] :: u_5613_1 []
    ) :: u_5613_2
  );
```

And then simply calling:

```java
Result<V_5613Record> result = ctx.fetch(V_5613);
```

I'm getting this error:

```
org.jooq.exception.DataTypeException: Error while creating array
	at org.jooq.impl.DefaultBinding.pgNewArray(DefaultBinding.java:2322)
	at org.jooq.impl.DefaultBinding.pgFromString(DefaultBinding.java:2170)
	at org.jooq.impl.DefaultBinding.pgSetValue(DefaultBinding.java:2327)
	at org.jooq.impl.DefaultBinding$2.operate(DefaultBinding.java:2220)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:128)
	at org.jooq.impl.DefaultBinding.pgNewRecord(DefaultBinding.java:2212)
	at org.jooq.impl.DefaultBinding.get(DefaultBinding.java:1521)
	at org.jooq.impl.DefaultBinding.pgGetArray(DefaultBinding.java:2259)
	at org.jooq.impl.DefaultBinding.get(DefaultBinding.java:1499)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1653)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1624)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:128)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchOne(CursorImpl.java:1553)
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1520)
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:342)
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:295)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:282)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:349)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:312)
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2708)
	at org.jooq.impl.DefaultDSLContext.fetch(DefaultDSLContext.java:3662)
	at org.jooq.impl.DefaultDSLContext.fetch(DefaultDSLContext.java:3657)
```

Does that seem to be the same issue as yours?

**Comment by lukaseder** [comment]
This issue looks similar: https://github.com/jOOQ/jOOQ/pull/5586. Will check if that's the same problem.

**Comment by BrentDouglas** [comment]
@lukaseder Yes, they are the same issue

**Comment by robertotru** [comment]
Hi Lukas.

Yes, the stacktrace is the same.

Best
Roberto

**Comment by lukaseder** [comment]
Fixed this for jOOQ 3.9, will backport to 3.8.6. With my fix, `v_5613` from my previous comment (https://github.com/jOOQ/jOOQ/issues/5613#issuecomment-255401791) can now be correctly loaded. I will be rejecting the suggested PR by @BrentDouglas (https://github.com/jOOQ/jOOQ/pull/5586) as it changes a lot of code when the fix can be done much more easily. If I misunderstood the PR, we can continue discussing directly on #5586.
